sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/RegistrationKeyManager.java

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.android.communication.chat.implementation.notifications.fcm;
+
+import android.util.Log;
+
+import com.azure.android.core.logging.ClientLogger;
+import com.azure.android.core.util.Base64Util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.security.NoSuchAlgorithmException;
+import java.util.Stack;
+
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+
+/**
+ * A singleton wrapper which manage the persistence and rotation of secrete keys and execution result related to #{RenewalTokenWorker}.
+ * The secrete keys are persisted and loaded using key-store API. We generate new secrete keys every refresh and evict expired keys
+ * from key-store every refresh.
+ *
+ * RegistrationKeyStore is used as object to manage the keys in file/memory.
+ */
+public final class RegistrationKeyManager {
```

Comment:
As a general note, we should have run approach by the Crypto Board. The folks working on this feature for iOS were told the same and if I recall correctly, they were going to schedule a meeting with them. At first glance what you're doing seems right to me but it's better to check with the Board to ensure we are not putting customer data at risk.

Reply:
I think this is safe approach. We are persisting each credential for 45 minutes. And those credentials are not notification data. They are the secreteKey to decrypt the actual data. The push notification data itself are never being persisted in the device.

Reply:
We should still run this by the Crypto Board. You can send them an email directly or coordinate with the iOS folks if they already have a conversation going or set up a meeting.